src/extras/TextureUtils.js

```diff
@@ -0,0 +1,141 @@
+import { AlphaFormat, LuminanceFormat, LuminanceAlphaFormat, RedFormat, RedIntegerFormat, RGFormat, RGIntegerFormat, RGBFormat, RGBAFormat, RGBAIntegerFormat, RGB_S3TC_DXT1_Format, RGBA_S3TC_DXT1_Format, RGBA_S3TC_DXT3_Format, RGBA_S3TC_DXT5_Format, RGB_PVRTC_2BPPV1_Format, RGBA_PVRTC_2BPPV1_Format, RGB_PVRTC_4BPPV1_Format, RGBA_PVRTC_4BPPV1_Format, RGB_ETC1_Format, RGB_ETC2_Format, RGBA_ETC2_EAC_Format, RGBA_ASTC_4x4_Format, RGBA_ASTC_5x4_Format, RGBA_ASTC_5x5_Format, RGBA_ASTC_6x5_Format, RGBA_ASTC_6x6_Format, RGBA_ASTC_8x5_Format, RGBA_ASTC_8x6_Format, RGBA_ASTC_8x8_Format, RGBA_ASTC_10x5_Format, RGBA_ASTC_10x6_Format, RGBA_ASTC_10x8_Format, RGBA_ASTC_10x10_Format, RGBA_ASTC_12x10_Format, RGBA_ASTC_12x12_Format, RGBA_BPTC_Format, RGB_BPTC_SIGNED_Format, RGB_BPTC_UNSIGNED_Format, RED_RGTC1_Format, SIGNED_RED_RGTC1_Format, RED_GREEN_RGTC2_Format, SIGNED_RED_GREEN_RGTC2_Format, UnsignedByteType, ByteType, UnsignedShortType, ShortType, HalfFloatType, UnsignedShort4444Type, UnsignedShort5551Type, UnsignedIntType, IntType, FloatType, UnsignedInt5999Type } from '../constants.js';
+
+/**
+ * Given the width, height, format, and type of a texture. Determines how many
+ * bytes must be used to represent the texture.
+ */
+function getByteLength( width, height, format, type ) {
```

Comment:
Can this function be internal or is it required for the user to have access to it?

If possible, I would prefer to not expose it and move the logic to `WebGLUtils` instead. In a different PR, there is the plan to add `TextureUtils` as an addon in `examples/jsm` (see https://github.com/mrdoob/three.js/pull/28512#issuecomment-2139001685).

Reply:
Oh wait, I see `getByteLength()` is used in the new example...

Reply:
Since this is an advanced feature, I would probably not expose the helper for now and inline the size computation in the example.

Reply:
Hmm, I think `addLayerUpdate` has limited value if three doesn't provide clients a way to know the byte bounds of a single layer. You need `TextureUtils.getByteLength` to know which subarray within the `CompressedArrayTexture`'s data to update. We could make this method internal, but then it would force clients to either implement this logic themselves or copy/paste it from the three repo.

Reply:
And inlining would fail in the example because the KTX2Loader dynamically selects a format + type depending on the underlying WebGL context. There's no guarantee that the hard-coded byte length will be consistent on different devices.

Reply:
(also thanks for the quick review)

Reply:
Let's see how we can organize the code regarding #28512.

Reply:
The code in https://github.com/mrdoob/three.js/pull/28512 is so compact so I think we can move it in `src` as well. That means there is no need to update this PR.

It's definitely good that the layer size computation in `uploadTexture()` is moved into a helper function.